src/fontra/views/editor/scene-controller.js

```diff
@@ -1333,6 +1371,70 @@ function getSelectedContours(path, pointSelection) {
   return [...selectedContours];
 }
 
+function getSelectedOpenContours(
+  path,
+  pointSelection,
+  ignoreSinglePointContour = true
+) {
+  if (!path || !pointSelection) {
+    return [];
+  }
+  const selectedContours = new Set();
+  for (const contourIndex of getSelectedContours(path, pointSelection)) {
+    if (path.contourInfo[contourIndex].isClosed) {
+      // skip if contour is closed already
+      continue;
+    }
+    const contour = path.getContour(contourIndex);
+    const numOnCurvePoints = contour.pointTypes.reduce(
+      (acc, pointType) =>
+        acc +
+        ((pointType & VarPackedPath.POINT_TYPE_MASK) === VarPackedPath.ON_CURVE
+          ? 1
+          : 0),
+      0
+    );
+    if (numOnCurvePoints === 1 && ignoreSinglePointContour) {
+      // skip single point contour
+      continue;
+    }
+    selectedContours.add(contourIndex);
+  }
+
+  return [...selectedContours];
+}
+
+function closeContourEnsureCubicOffCurves(path, contourIndex) {
+  const startPoint = path.getContourPoint(contourIndex, 0);
+  const secondPoint = path.getContourPoint(contourIndex, 1);
+  const prevEndPoint = path.getContourPoint(contourIndex, -2);
+  const endPoint = path.getContourPoint(contourIndex, -1);
+
+  const offCurveAtStart = !secondPoint.type && startPoint.type && !endPoint.type;
+  const firstPoint = offCurveAtStart ? secondPoint : prevEndPoint;
+  const middlePoint = offCurveAtStart ? startPoint : endPoint;
+  const lastPoint = offCurveAtStart ? endPoint : startPoint;
+
+  if (firstPoint.type || !middlePoint.type || lastPoint.type) {
+    // Sanity check: we expect on-curve/off-curve/on-curve
+    return;
+  }
+
+  // Compute handles for a cubic segment that will look the same as the
+  // one-off-curve quad segment we have.
+  const [handle1, handle2] = [firstPoint, lastPoint].map((point) => {
+    return {
+      ...vector.roundVector(scalePoint(point, middlePoint, 2 / 3)),
+      type: "cubic",
+    };
+  });
+
+  const numContourPoints = path.getNumPointsOfContour(contourIndex);
+  const handle1ContourIndex = offCurveAtStart ? 0 : numContourPoints - 1;
```

Comment:
What you call `handle1ContourIndex` is in fact a contour point index.

Btw. setContourPoint and insertPoint also take negative point indices, so I'm 99% sure you don't need the numContourPoints here.

Ah, but you want to append at the end of the contour, which won't work with a negative index. However, there is path.appendPoint().

Reply:
Thank you very much for your caring feedback. I removed numContourPoints and used path.appendPoint().